Add support for multi-index query parameters for `_cluster/state`

cbuescher opened the pull request:
Added missing support for the multi-index query parameters 'ignore_unavailable',
'allow_no_indices', and 'expand_wildcards' to '_cluster/state' API. These
parameters are supposed to be supported for APIs that refer to index parameter
support execution across multiple indices.

Closes #5229

Commit:
Rest: Adding support of multi-index query parameters for _cluster/state

Added missing support for the multi-index query parameters 'ignore_unavailable',
'allow_no_indices', and 'expand_wildcards' to '_cluster/state' API. These
parameters are supposed to be supported for APIs that refer to index parameter
support execution across multiple indices.

Closes #5229

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateRequest.java:
```diff
@@ -127,6 +133,7 @@ public void readFrom(StreamInput in) throws IOException {
         metaData = in.readBoolean();
         blocks = in.readBoolean();
         indices = in.readStringArray();
+        indicesOptions = IndicesOptions.readIndicesOptions(in);
```
we need to add a version check here on the 1.x branch, to make sure that we maintain bw compatibility:
```
if (in.getVersion().onOrAfter(Version.1_5_0)) {
    indicesOptions = IndicesOptions.readIndicesOptions(in);
}
```

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateRequest.java:
```diff
@@ -137,5 +144,6 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeBoolean(metaData);
         out.writeBoolean(blocks);
         out.writeStringArray(indices);
+        indicesOptions.writeIndicesOptions(out);
```
we need to add a version check here on the 1.x branch, to make sure that we maintain bw compatibility:
```
if (out.getVersion().onOrAfter(Version.1_5_0)) {
    indicesOptions = IndicesOptions.writeIndicesOptions(out);
}
```

javanna commented on rest-api-spec/test/cluster.state/40_allow_no_indices.yaml:
```diff
@@ -0,0 +1,18 @@
+        index: test*
+        allow_no_indices: false
+
+  - is_false: metadata
```
I think this assert doesn't make much sense given that the last call returns an error. `catch: missing` seems enough.

javanna commented on rest-api-spec/test/cluster.state/30_expand_wildcards.yaml:
```diff
@@ -0,0 +1,77 @@
+        index: foobla
+        ignore_unavailable: false
+
+  - is_false: metadata
```
same as below, I would remove this assertion

javanna commented on rest-api-spec/test/cluster.state/40_allow_no_indices.yaml:
```diff
@@ -0,0 +1,18 @@
+---
```
I was wondering if it makes sense to merge this file with the previous one, since they both test indices options?

javanna commented:
Left a few comments, looks good though!

Commit:
Changes according to review comments

Commit:
adding tests for different settings of indices options on ClusterStateRequest

cbuescher commented:
Added changes according to your comments and also added tests on the ClusterStateRequest. Is SimpleClusterStateTests an okay place for such integration tests or is there a better place?

As for the Versions and Serialization tests I would need some more feedback, will message you about that soon.

javanna commented on src/test/java/org/elasticsearch/cluster/SimpleClusterStateTests.java:
```diff
@@ -157,4 +160,51 @@ public void testLargeClusterStatePublishing() throws Exception {
+        IndicesOptions ignoreUnavailabe = IndicesOptions.fromOptions(true, true, true, false);
+        clusterStateResponse = client().admin().cluster().prepareState().clear().setMetaData(true).setIndices("fzzbzz")
+                .setIndicesOptions(ignoreUnavailabe).get();
+        assertTrue(clusterStateResponse.getState().metaData().indices().isEmpty());
```
for consistency: assertThat(clusterStateResponse.getState().metaData().indices().isEmpty(), is(true));

javanna commented on src/test/java/org/elasticsearch/cluster/SimpleClusterStateTests.java:
```diff
@@ -157,4 +160,51 @@ public void testLargeClusterStatePublishing() throws Exception {
+        IndicesOptions allowNoIndices = IndicesOptions.fromOptions(false, true, true, false);
+        clusterStateResponse = client().admin().cluster().prepareState().clear().setMetaData(true).setIndices("a*")
+                .setIndicesOptions(allowNoIndices).get();
+        assertTrue(clusterStateResponse.getState().metaData().indices().isEmpty());
```
same as above

javanna commented:
Hey @cbuescher `SimpleClusterStateTests` looks like a good place for those tests, that's fine. As for tests around version checks, have a look here for instance: https://github.com/elasticsearch/elasticsearch/blob/1.x/src/test/java/org/elasticsearch/action/support/IndicesOptionsTests.java#L33 .

Commit:
Added tests for ClusterStateRequest serialization

Commit:
Adding serialization test for ClusterStateRequest

cbuescher commented:
I added serialization test in https://github.com/cbuescher/elasticsearch/blob/fix/5229/src/test/java/org/elasticsearch/action/admin/cluster/state/ClusterStateRequestTest.java for master. 
I think I will open a second PR for the changes in 1.x and include the right version checks and tests for that there.

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/state/ClusterStateRequestTest.java:
```diff
@@ -0,0 +1,65 @@
+            ClusterStateRequest clusterStateRequest = new ClusterStateRequest().routingTable(randomBoolean()).metaData(randomBoolean())
+                    .nodes(randomBoolean()).blocks(randomBoolean()).indices("testindex", "testindex2").indicesOptions(indicesOptions);
+
+            BytesStreamOutput output = new BytesStreamOutput();
```
I think you want to randomize version of out and in to make sure that we keep bw comp. In master that makes only little sense now but it will be good in the future. I think we can ranzomize versions starting from the minCompatibilityVersion on. That should make the test the same in both master and 1.x

javanna commented on src/test/java/org/elasticsearch/cluster/SimpleClusterStateTests.java:
```diff
@@ -19,15 +19,19 @@
 
 package org.elasticsearch.cluster;
 
+import org.apache.tools.ant.taskdefs.condition.IsTrue;
```
this is the wrong import ;)

Commit:
adding randomization to test

Commit:
adding ClusterStateRequest backwards compatibility test

cbuescher commented:
I just added backwards compatibility test to basically check that nothing breaks when introducing the new IndicesOptions field on the 1.x branch. Already tested that on 1.x, will merge that there. Is it possible to use this test on master as well? I think it is disabled by defautl, right?

javanna commented on src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompat.java:
```diff
@@ -0,0 +1,56 @@
+import java.lang.reflect.Method;
+
+
+@ElasticsearchIntegrationTest.ClusterScope(scope= ElasticsearchIntegrationTest.Scope.SUITE,  numClientNodes = 0)
```
why did you use scop SUITE here? and why no client nodes?

cbuescher replied:
Think I got that from NodesStatsBasicBackwardsCompat test, no reason besides that. Should I change that, and if yes, whats the prefered setting here?

javanna replied:
remove the whole annotation, it inherits from ElasticsearchBackwardsCompatIntegrationTest that has already scope SUITE and numClientNodes set to 0, defaults should be good, no need to override them

Commit:
use default settings in ClusterStateBackwardsCompat test
improving on backwards test

javanna commented on src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompat.java:
```diff
@@ -0,0 +1,56 @@
+
+        // connect to each node with a custom TransportClient, issue a ClusterStateRequest to test serialization
+        for (NodeInfo n : nodesInfo.getNodes()) {
+            TransportClient tc = new TransportClient(settings).addTransportAddress(n.getNode().address());
```
this needs to be wrapped into a try finally block, so no explicit `close` is needed.

javanna commented on src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompat.java:
```diff
@@ -0,0 +1,56 @@
+public class ClusterStateBackwardsCompat extends ElasticsearchBackwardsCompatIntegrationTest {
+
+    @Test
+    public void testClusterState() throws Exception {
```
given the randomized behaviour of client() I am wondering if it makes sense to go and test the call from all the potential nodes. That happens already eventually with a single call that uses client() and no loop, no nodes info call either. Unless I am missing something.

cbuescher replied:
I didn't knwo about random client(), just saw this. However, if I do this, how can I be sure that I test serialization of request between both potential versions? As far as I understand it so far you can be lucky an just always hit the nodes with the newer versions, or am I mistaken? If by "eventually" you mean after running this test lots of times, that will be okay I guess. Are these tests run multiple times on the test machines?

javanna replied:
you got it, by eventually I mean after running the test some times (I don't think it's going to be a lot of times given that we have only a few nodes in this bw setup). These tests run consitnuosly on our CI. I think we are good if we simplify the test a bit as I said above. Makes sense?

javanna commented:
Great work @cbuescher I left a couple of comments on the bw comp test but it looks good, this is very close. As for the bw comp test, you can push to master as well, no problem.

cbuescher commented:
@javanna I slightly modified the backwards compat test but kept the loop over the nodes in there. Did some logging and with the random client() it seems that the request not always runs over nodes in all versions present. Not sure why, do you think this is too complicated as a test now? Any other idea would be welcome.

Commit:
corrected comments

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/state/ClusterStateRequestTest.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    private void assertOptionsMatch(IndicesOptions in, IndicesOptions out) {
```
nitpick, this method could be static

javanna commented on src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompat.java:
```diff
@@ -0,0 +1,54 @@
+        // ClusterStateRequest to test serialization
+        for (NodeInfo n : nodesInfo.getNodes()) {
+            TransportClient tc = null;
+            try {
```
sorry my previous comment wasn't clear. I meant this:
```
try (TransportClient tc = new TransportClient(settings)) {
    tc.addTransportAddress(n.getNode().address());
    tc.admin().cluster().prepareState().clear().execute().actionGet();
}
```

javanna commented on src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompat.java:
```diff
@@ -0,0 +1,54 @@
+            TransportClient tc = null;
+            try {
+                tc = new TransportClient(settings).addTransportAddress(n.getNode().address());
+                tc.admin().cluster().prepareState().clear().execute().actionGet();
```
maybe also look at what it returns and run some assertions on the response?

cbuescher replied:
I tried to make some reasonable assertions on the response but found that with the current test setup there is not much interesting to assert. Only made sure that the name and state are set in the response now.

Commit:
corrected try-with-resource statement, added assertions

Commit:
adding to bwcompat test

cbuescher commented:
@javanna changed the tests according to your suggestions, looks okay now?

javanna commented:
LGTM
